Feature: play easier together with friends from behind home routers

TrueBrain opened the pull request:
Integrated #9234. Closes #9234.

~Test binaries can be found here: https://www.openttd.org/downloads/openttd-branches/pr9017/latest.html
See below how to help out testing.~

~If you build yourself, make sure to name your branch "pr9017". Otherwise your revision will not match anyone using the binaries above.~

**!!NOTE!!** Currently the Game Coordinator is not working with this PR, while I rebuild this PR from the ground up. See https://gist.github.com/TrueBrain/75c016cdb90efa18cb79f3f6bae9af25 for details.

## Motivation / Problem

Currently, to host a network server you really need to know a few things about routers and networking. This is way to advanced for the average Joe, and this becomes very clear when looking at Steam Discussions.
For Rich Presence (Discord, Steam, ..) we need "invite codes" so people can join each others game.

So, can we do this? I wrote a gist to explain the process of getting to this PR: https://gist.github.com/TrueBrain/f5685dc0fdbd012aac48ea2a36158198

But in case you are not into reading, let me summarize:

- Every invite-only / public server creates a persistent TCP connection to a Game Coordinator (replacing the master-server)
- Every server gets assigned an unique "join key" (called an "invite code" in the UI)
- Server listing returns only the "join key" of a server
- A client can join any server based on this "join key" (without even knowing the IP of the server), either via listing or directly inputting it
- The Game Coordinator will try to connect the server and client. It tries 3 methods: direct connect, STUN, and if all fails, TURN (see gists for details)

There are now three kind of servers:
- Local servers: do not connect to the Game Coordinator. "Search LAN" does work for these servers, as do people with the public IP and if port-forwarding is setup correctly. Just the old "non-advertised" servers.
- Invite-only servers: the owner of the server needs to give the invite-code to someone before he can join.
- Public servers are the old advertised servers.

## Description

See the individual commits and the gist above. Not much more to add.

## Limitations

- Server info is delayed by 30s (by design). You can refresh individual servers if you want up-to-date information.

## Testing

If you want to help out testing, great!

- Start the client by downloading it here: https://www.openttd.org/downloads/openttd-branches/pr9017/latest.html (or build it yourself, make sure you name your branch `pr9017`!)
- Start a server, make it public or invite-only
- Ask someone to join your server

Notice how you didn't have to setup firewalls, port-forwarding, etc?

Let me know if you find any problems, etc.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TheDude-gh committed:
Add: inform clients what game-script a server is running

nielsmh commented on src/network/network_gui.cpp:
```diff
@@ -233,6 +234,8 @@ class NetworkGameWindow : public Window {
 	int blot_offset; ///< Left offset for green/yellow/red compatibility icon.
 	int flag_offset; ///< Left offset for language flag icon.
 
+	byte widget_id; ///< The widget that has the pop-up input menu.
```
This could be named better. Suggesting `query_owner` although that's a bit long.
Really maybe `ShowQueryString` should be changed to take a std::function instead of an object.

TrueBrain replied:
I just copied what the other windows already were doing :D But I do agree ;)

nielsmh reviewed (commented):
I didn't read everything in detail, but this doesn't remove local network UDP broadcast query, right?

TrueBrain commented:
> 
> 
> I didn't read everything in detail, but this doesn't remove local network UDP broadcast query, right?

Local broadcast will remain, yes.

James103 commented:
> * Load test Game Controller if it can handle 2k servers and 5k clients

How exactly did you arrive at the load test goals (2,000 servers and 5,000 clients)?

TrueBrain commented:
> 
> 
> > * Load test Game Controller if it can handle 2k servers and 5k clients
> 
> How exactly did you arrive at the load test goals (2,000 servers and 5,000 clients)?

I was sure I wrote that down somewhere .. guess I did not :D

It is a worst-case estimate: on peak on Steam we had 3.2k people online. Say, with a good multiplayer, they all want to play with their friends, round it up, gives an upper value of ~5k.
Assuming most people will just play with their friends, means ~2.5 people per server, so 2k servers.

Infrastructure wise, being able to handle that, means I can sleep peaceful at night.

You think I am understating? Let me know :)

TrueBrain committed:
Add: support STUN for network connections

TrueBrain committed:
Codechange: allow Connect() to bind to a local address

TrueBrain committed:
Codechange: track servers with a ServerAddress instead of a NetworkAddress

This allows future extensions to have different ways of referencing
a server, instead of forcing to use IP:port.

TrueBrain committed:
Add: use Game Coordinator to annouce public servers

TrueBrain committed:
Remove: old server announcement to Master Server

The Master Server doesn't understand NetworkGameInfo v5, so it is
of no use anyway.

TrueBrain committed:
Add: use Game Coordinator to get latest public server listing

TrueBrain committed:
Codechange: track version of network servers to prune once out-of-date

TrueBrain committed:
Remove: old server listing via Master Server

This removes all UDP from the game except for a local broadcast
to find LAN games.

TrueBrain committed:
Add: allow connecting to server based on join-key

TrueBrain committed:
Add: support TURN for network connections

TrueBrain committed:
Add: allow setting your server visibility to "invite-only"

In this mode you do register to the Game Coordinator, but your
server will not show up in the public server listing. You can give
your friends the invite code of the server with which they can
join.

rubidium42 commented on src/network/core/tcp_connect.cpp:
```diff
@@ -36,7 +47,24 @@ TCPConnecter::TCPConnecter(const NetworkAddress &address) :
+	this->address = address;
+
+	_tcp_connecters.push_back(this);
+	if (!StartNewThread(nullptr, "ottd:tcp", &TCPConnecter::ThreadEntry, this)) {
```
I would use a different thread name here, so it can be distinguished later on.

TrueBrain replied:
Not sure they is a good idea; they are in fact all the same thing. But I think it is better to move this code into a single function that they all three call. That way it is far more obvious that it is the same thing.

In general, this creating of a thread in the ctor has been a true headache .. but okay .. another story :D

rubidium42 commented on src/network/core/tcp_connect.cpp:
```diff
@@ -64,6 +64,32 @@ TCPBindConnecter::TCPBindConnecter(const NetworkAddress &address, NetworkAddress
+		case SERVER_ADDRESS_DIRECT:
+			this->address = static_cast<const ServerAddressDirect *>(server_address.get())->address;
+
+			if (!StartNewThread(nullptr, "ottd:tcp", &TCPConnecter::ThreadEntry, this)) {
```
Different thread, different name.

rubidium42 commented on src/network/core/address.h:
```diff
@@ -183,4 +183,97 @@ class NetworkAddress {
+	/**
+	 * Create a copy of the object.
+	 */
+	virtual std::unique_ptr<ServerAddress> Copy() = 0;
```
Do you really need a copy, or is maybe a std::shared_ptr enough?

rubidium42 commented on src/network/network_coordinator.cpp:
```diff
@@ -66,6 +67,40 @@ bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_REGISTER_ACK(Packet *
+		/* Clear any existing GRFConfig chain. */
+		ClearGRFConfigList(&item->info.grfconfig);
+		/* Copy the new NetworkGameInfo info. */
+		memcpy(&item->info, &ngi, sizeof(item->info));
```
This could be just `item->info = ngi` as far as I'm aware.

rubidium42 commented on src/network/network_gamelist.cpp:
```diff
@@ -179,6 +179,8 @@ void NetworkGameListRequery()
 				NetworkUDPQueryServer(static_cast<ServerAddressDirect *>(item->address.get())->address);
 				break;
 
+			/* Will never support SERVER_ADDRESS_JOIN_KEY. */
```
I would make it `case SERVER_ADDRESS_JOIN_KEY: // Will never be supported` so it becomes slightly more apparent in the IDE instead of just a comment.

TrueBrain replied:
It was more a mental note, as that function gets removed a few commits later :)

rubidium42 commented on src/network/network_gui.cpp:
```diff
@@ -756,6 +760,16 @@ class NetworkGameWindow : public Window {
+				this->widget_id = WID_NG_JOIN_KEY;
+				SetDParamStr(0, "");
+				ShowQueryString(
+					STR_JUST_RAW_STRING,
```
Why not `STR_EMPTY` instead of raw string + setting a string parameter?

rubidium42 commented on src/network/core/tcp_coordinator.cpp:
```diff
@@ -48,6 +48,7 @@ bool NetworkCoordinatorSocketHandler::HandlePacket(Packet *p)
 		case PACKET_COORDINATOR_SERVER_DIRECT_CONNECT: return this->Receive_SERVER_DIRECT_CONNECT(p);
 		case PACKET_COORDINATOR_SERVER_STUN_REQUEST:   return this->Receive_SERVER_STUN_REQUEST(p);
 		case PACKET_COORDINATOR_SERVER_STUN_CONNECT:   return this->Receive_SERVER_STUN_CONNECT(p);
+		case PACKET_COORDINATOR_SERVER_STUN_ACCEPT:   return this->Receive_SERVER_STUN_ACCEPT(p);
```
Minor misalignment sneaked in.

rubidium42 commented on src/network/network_coordinator.h:
```diff
@@ -0,0 +1,93 @@
+protected:
+	bool Receive_SERVER_REGISTER_ACK(Packet *p) override;
+	bool Receive_SERVER_LISTING(Packet *p) override;
+	bool Receive_SERVER_CONNECTING(Packet *p);
```
Missing `override` (thus some warnings).

TrueBrain committed:
Remove: unused UDP protocol parts

TrueBrain committed:
Cleanup: remove unused optional ctor parameter for NetworkGameList

TrueBrain committed:
Change: multiple modes to send NewGRF data to client / Game Coordinator

NewGRFs cannot be changed in multiplayer games, and consumes a lot
of bytes. So only send it once to the Game Coordiantor on registration,
and skip the field after that for any update.
A new game will restart the registration, so any NewGRF change
between games is picked up via the new registration.

Clients that look at the NewGRFs a server has, want to know the names
of the NewGRFs more than the IDs. So in addition to the ID and MD5,
send the full name to the clients too.

This solution is picked so the Game Coordinator isn't sending the
names of all the NewGRFs to the clients constantly (which would
make the server-listing very large), but still allow clinets to
see the name of a specific server if he clicks the button.

TrueBrain commented:
This is pretty much done and ready for review. I am 100% sure there are bugs, so lets find them and fix them :)

The Game Coordinator is not yet running on AWS, and is not scalable yet. Otherwise it should do its job. So any client-facing bug you can find, please report that.

My suggestion is to get this PR code-reviewed and merged. After that I will work on the Game Coordinator, to make it production-ready. But that also gives time to test the client itself.

There is plenty of future-work to be done, but this can be found back here:
https://gist.github.com/TrueBrain/d430a5f0e7e4312b895278f092067495

rubidium42 commented on src/network/core/address.cpp:
```diff
@@ -421,3 +422,7 @@ void NetworkAddress::Listen(int socktype, SocketList *sockets)
 	return NetworkAddress(addr, addr_len).GetAddressAsString();
 }
+
+ServerAddressDirect::ServerAddressDirect(const std::string &hostname, uint16 port) : ServerAddress(SERVER_ADDRESS_DIRECT) {
```
Missing newline

rubidium42 commented on src/network/network_gamelist.cpp:
```diff
@@ -121,6 +123,37 @@ void NetworkGameListRemoveItem(NetworkGameList *remove)
+
+			/* Remove GRFConfig information */
+			ClearGRFConfigList(&remove->info.grfconfig);
+			free(remove);
```
NetworkGameList are created with `new`, so you have to use `delete` here.

rubidium42 commented on src/network/network_gamelist.cpp:
```diff
@@ -121,6 +123,37 @@ void NetworkGameListRemoveItem(NetworkGameList *remove)
+ */
+void NetworkGameListRemoveExpired()
+{
+	NetworkGameList *prev_item = nullptr;
```
What you could do is make this `NetworkGameList **prev_item = &_network_game_list` and at line 149 `prev_item = &item->next`, then lines 139-143 can become `*prev_item = item;`.
Practically you could even remove item completely, if you replace every instance of `item` with `(*prev_item)` though that make things less legible.
It's basically what is used in some places in the packet code queue code.

rubidium42 commented on src/network/network.cpp:
```diff
@@ -542,6 +555,11 @@ NetworkAddress ParseConnectionString(const std::string &connection_string, uint1
 {
+	if (!connection_string.empty() && connection_string[0] == '+') {
+		std::string_view ip = ParseCompanyFromConnectionString(connection_string, company);
+		return std::make_unique<ServerAddressJoinKey>(std::string(ip.substr(1)));
```
Given I'm seeing the `+` getting added in many places, and it getting removed here, I am wondering why we can't just send it with a `+` over the network. Or, at least once we received it immediately add the `+` to it and only remove it when we send it again.
It would also mean that that ServerDirectAddress and ServerJoinKey classes are basically identical in functionality; it's a type plus a string, so that might remove the need for the specific sub classes and associated casts, and other complications like unique_ptr.

TrueBrain replied:
I didn't want to send `+` over the network, as this is purely a client-only way method to represent join-keys. The benefit is that it means we can change it to what-ever when-ever we like.

But yeah, the constant adding and removing is a bit annoying; I did try to add/remove it on network level only, but that also wasn't the prettiest code. I will give it another look soon.

TrueBrain replied:
Did some heavy rewriting on this part, not only with the `+`, but how `ServerAddress` works in general. I think it is a lot cleaner now :)

rubidium42 commented on src/network/network_coordinator.cpp:
```diff
@@ -156,6 +198,45 @@ bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_LISTING(Packet *p)
+
+	/* Find the connecter based on the join-key. */
+	auto connecter_it = this->connecter_pre.find(join_key);
+	assert(connecter_it != this->connecter_pre.end());
```
I'm not happy with the location of this assert, and any other in a similar location. Given it comes from the network the content of `Packet` should be considered invalid, i.e. it will trigger the assertion (or not in a release build, resulting in undefined behavior). So, it should either just ignore the Packet if it is with some comments or kill the connection with some error message.

TrueBrain replied:
I thought I fixed them all; clearly missed one.

This btw is the only one, as far as I can see. The others are not in response to an incoming packet. They already should have validated that their token/join-key is valid, and as such an assert is perfectly fine I think?

rubidium42 commented on src/network/network_coordinator.cpp:
```diff
@@ -236,6 +325,72 @@ void ClientNetworkCoordinatorSocketHandler::GetListing()
+
+	Packet *p = new Packet(PACKET_COORDINATOR_CLIENT_CONNECT);
+	p->Send_uint8(NETWORK_COORDINATOR_VERSION);
+	p->Send_string(join_key.c_str());
```
I have changed Send_string so it should not be needed to `c_str()` it.

rubidium42 commented on src/network/core/address.cpp:
```diff
@@ -326,13 +326,10 @@ static SOCKET ListenLoopProc(addrinfo *runp)
-	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, (const char*)&on, sizeof(on)) == -1) {
-		DEBUG(net, 3, "[%s] could not set reusable %s sockets for port %s: %s", type, family, address.c_str(), NetworkError::GetLast().AsString());
-	}
+	if (!SetReusePort(sock)) DEBUG(net, 0, "[%s] setting reuse-port mode failed for %s port %s", type, family, address.c_str());
```
Since you're touching this, the second %s says "IPvX" and the last %s gives you "IP:port (IPvX)" so that's duplicate information. Similarly "IP:port" is not really a port, is it? And is it wanted that the error from the OS is not passed to the "user" in this situation, as that might be the only thing that can help us in any way.

TrueBrain replied:
To not bloat this PR any more than I have to, #9230 :)

rubidium42 commented on src/network/core/tcp_stun.h:
```diff
@@ -0,0 +1,60 @@
+	{
+	}
+
+	/** On destructing of this class, the socket needs to be closed */
```
Missing `.` at the end of the sentence? Also with the constructor.

rubidium42 commented on src/network/network_coordinator.cpp:
```diff
@@ -208,8 +240,8 @@ bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_CONNECTING(Packet *p)
 	/* Now store it based on the token. */
-	this->connecter_pre.erase(connecter_it);
 	this->connecter[token] = connecter_it->second;
+	this->connecter_pre.erase(connecter_it);
```
In commit 'Add: support STUN for network connections' you move this line by one, should it be put in the right place in the commit that added these lines?

rubidium42 commented on src/network/core/game_info.cpp:
```diff
@@ -264,27 +274,35 @@ void DeserializeNetworkGameInfo(Packet *p, NetworkGameInfo *info)
+					GRFConfig *c = new GRFConfig();
+					DeserializeGRFIdentifier(p, &c->ident);
+					std::string name = {};
+					if (newgrf_mode == GAME_INFO_NEWGRF_MODE_FULL) {
```
Do you know the average number of times the same NewGRF is used by different servers? Like NewGRF A is used by 10 servers, NewGRF B by 5, NewGRF C by 1, so the average is about 16/5 ~= 5. That could be used to influence choosing an even more efficient way.
You would then be creating a list of used NewGRFs at the top of the packet with the whole listing, and here you pass indices into that list. Those indices can probably be uint16 but stick with uint32 for now for an added safety buffer. At an average of 2 you would be saving about 16 + NewGRF name length bytes per NewGRF, with an average at 3 you would save about 32 + 2 * NewGRF name length, and so forth.

TrueBrain replied:
There are tons of more efficient ways honestly. I liked the suggestion to apply compression on such packets .. that also sound really useful, and maybe means we don't even need a trick like yours. But that needs more work and benchmarking.

For now, I went for a solution that was simple and elegant, rather than complicated and touching a lot of stuff. Future work could optimize this by a lot. But I rather not do that in this PR :) If you agree, ofc.

Btw, this last commit is purely here as otherwise after this PR we would have lost functionality we had before the PR.

TrueBrain committed:
Feature: raise the maximum NewGRF limit to 255

glx22 commented:
Did some testing with a dedicated server and a client.
- connection type is "Behind NAT", while a build from master is correctly advertised on master server, so direct connection should work.
- clicked "search internet", server is listed, its address is the invite code
- clicked "search LAN", server is added with one of the local ipv4
- clicked "search LAN" again, server is added with the other local ipv4
- after a short time both local ipv4 are removed from the list, and I can see some redrawing issues for the server list too
- joined the only left server as spectator, works as expected

I had an issue with a ":3979" server, but I can't reproduce it :(

Edit: I had the IPv6 firewall activated on the router, and it has only 2 modes on or off, no fine tuning. Disabled it to test and connection type became "Public". Build from master was advertising only the IPv4, as IPv6 was blocked by this "better than nothing" firewall.
So to resume:
- if server is IPv6 only, connection type is "Behind NAT" if IPv6 firewall is on, "Public" otherwise
- if server is IPv4 only, connection type is "Behind NAT" but should be "Public" because direct connection works (can advertise)
- if server is IPv4/IPv6, connection type is "Behind NAT" if IPv6 firewall is on, "Public" otherwise

TrueBrain commented:
>     * connection type is "Behind NAT", while a build from master is correctly advertised on master server, so direct connection should work.

Yeah, the GC only validates it for the IP you connect with (so IPv6 in your case; the IPv4 won't be detected atm). I will see what I can do about it :) EDIT: fixed!

>     * clicked "search internet", server is listed, its address is the invite code
>     * clicked "search LAN", server is added with one of the local ipv4
>     * clicked "search LAN" again, server is added with the other local ipv4
>     * after a short time both local ipv4 are removed from the list, and I can see some redrawing issues for the server list too

Yeah, fun bug. LAN servers were not marked as "do not expire", and as such when the server listing via GC came in, it noticed the LAN server was stale (didn't get an update), so it assumed it would be offline. And that removes it :) Fixed now :)

Tnx for testing, much appreciated :)

TrueBrain committed:
Codechange: use UpdateNetworkGameWindow() over manually marking window dirty

embeddedt commented:
@TrueBrain Theoretically, if WebSocket support is added to the Game Coordinator, this would allow Emscripten clients to join any publicly-advertised multiplayer game, with the coordinator server relaying inbound WS packets to TCP, right?

TrueBrain commented:
> 
> 
> @TrueBrain Theoretically, if WebSocket support is added to the Game Coordinator, this would allow Emscripten clients to join any publicly-advertised multiplayer game, with the coordinator server relaying inbound WS packets to TCP, right?

Absolutely. In a bit more details:

The Game Coordinator would need WebSocket support, and it has to understand it is a WebSocket client. As it shouldn't be trying Direct-Connect or STUN requests, but fall back directly to TURN.

The TURN servers need to support WebSocket too. So on one end they have a WebSocket, on the other they have the other side, which might be IPv4 / IPv6 / WebSocket. And it has to relay between them.

For now, however, I am holding back on this, as it can be rather costly to play relay for everyone. We first need to feel the water a bit how that is going to look :)

But yes, absolutely possible.

embeddedt commented:
> For now, however, I am holding back on this, as it can be rather costly to play relay for everyone.

That makes sense. This is some cool stuff you are making. :+1:

Milek7 commented on src/network/core/tcp_connect.cpp:
```diff
@@ -48,6 +77,16 @@ TCPConnecter::~TCPConnecter()
 	freeaddrinfo(this->ai);
```
This is sometimes called with NULL argument, but freeaddrinfo specs doesn't explicitly allow it.
It crashes on Haiku, and from what I can tell it will also crash on OpenBSD.

TrueBrain replied:
Not completely sure why this is a comment on this PR; I guess that is also why it got unnoticed till now :) But a fix is pending in https://github.com/OpenTTD/OpenTTD/pull/9404, as that code hasn't been part of this PR for a while now :D

TrueBrain commented:
I added all the wishes presented to me. The "invite code" is now kept as long as your `secrets.cfg` has a `server_join_key` and a valid `server_join_key_secret` that comes with that. When you first register, those are both given to you. If on your next registration you give those again, and the secret is valid, you can use the join-key.

Coordinator-side I tuned how invite codes are generated; they should feel more random now, and always ~7 chars.

I integrated #9234 in here.

Tons of bug-fixes and other things people noticed .. I think this should be about it. Coordinator still needs a bit of work, but that doesn't have to block this PR. The coordinator (and STUN and TURN server) are running 24/7, and should be stable enough to handle the nightly load.

I did load-testing on the coordinator .. even after 10k clients he was not worried. ~100MB RAM usage. Little CPU. So we are good there.

James103 commented:
Are you planning on running the server through a service that offers good DDOS protection? There's always the risk that someone might try to DDOS the Game Controller, STUN server, and/or TURN server with data coming from what appears to be many more clients than what these servers are designed to handle, and that risk will only increase as the game gets more popular.

TrueBrain commented:
> 
> 
> Are you planning on running the server through a service that offers good DDOS protection? There's always the risk that someone might try to DDOS the Game Controller, STUN server, and/or TURN server with data coming from what appears to be many more clients than what these servers are designed to handle, and that risk will only increase as the game gets more popular.

The risk doesn't increase or decrease with this, as we currently already run services that are at risk here :) So in that sense, the landscape is unchanged.

All our service are either offered via AWS or via OVH; both mitigate DDOS traffic automatically. And we do get those alerts from time to time.

That said, there is always a risk someone attacks our services specific, and those attacks are unlikely to be picked up by the DDOS filters (as we use a fully custom protocol). For that we have some manual control to mitigate that.
As example, a few months back 1 server was going bananas, and generated a DOS on our master-server. Sentry detected this, informed me, and I could mitigate the issue pretty easily.

So to answer your question: do we plan to? no, we already do ;) So currently I am not too worried. But tnx for asking, I really appreciate that! And let me know if you see a scenario happening that would not be covered / puts us at risk!

spnda commented on src/settings.cpp:
```diff
@@ -1535,21 +1600,58 @@ static void GRFSaveConfig(IniFile *ini, const char *grpname, const GRFConfig *li
+	proc(generic_ini, _misc_settings,    "misc",  nullptr, only_startup);
 #if defined(_WIN32) && !defined(DEDICATED)
-	proc(ini, _win32_settings,   "win32", nullptr, only_startup);
+	proc(generic_ini, _win32_settings,   "win32", nullptr, only_startup);
```
The spacing before `misc` and `win32` is not applicable anymore, because the other calls have been moved to the for loops below.

spnda commented on src/network/core/tcp_coordinator.cpp:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Helper for logging receiving invalid packets.
+ * @param type The received packet type.
+ * @return Always false, as it's an error.
```
If it's always false, why not just make it a `void` and handle the error in functions that call this, as they know it'll be an error?

TrueBrain replied:
This is part of how the network code is designed. This function is returned via the functions below that, that need to know if the packet is handled correctly or not.

In general you notice here the age of the network code, but I kept it as similar as possible as the other tcp_NNN files :)

spnda commented on src/network/core/tcp_listen.h:
```diff
@@ -30,6 +30,42 @@ class TCPListenHandler {
+		/* Can we handle a new client? */
+		if (!Tsocket::AllowConnection()) {
+			/* no more clients allowed?
+				* Send to the client that we are full! */
```
Comment not formatted correctly.

spnda commented on src/network/core/address.h:
```diff
@@ -174,7 +175,42 @@ class NetworkAddress {
+	/**
+	 * Create a new ServerAddress object.
+	 *
+	 * Please use ServerAddress::Parse() instead of calling this directly.
```
As we don't seem to use `new ServerAddress` outside of `ServerAddress::Parse` anyway, why not make this constructor private?

spnda commented on src/network/core/config.h:
```diff
@@ -71,14 +78,25 @@ static const uint NETWORK_CONTENT_VERSION_LENGTH  =   16;         ///< The maxim
- * This limit is reached when PACKET_UDP_SERVER_RESPONSE reaches the maximum size of UDP_MTU bytes.
+ * This limit exists to avoid SERVER_INFO exceeding the maximum MTU. At the
+ * time of writing this limit is 32767 (TCP_MTU).
+ * In SERVER_INFO is NetworkGameInfo, which is 142 bytes + 100 per NewGRF
```
Grammar issue?

spnda commented on src/network/core/os_abstraction.cpp:
```diff
@@ -159,6 +159,23 @@ bool SetNoDelay(SOCKET d)
+bool SetReusePort(SOCKET d)
+{
+#ifdef _WIN32
+	/* Windows has no SO_REUSEPORT, but for our usecases SO_REUSEADDR does the same job. */
+	int reuse_port = 1;
+	return setsockopt(d, SOL_SOCKET, SO_REUSEADDR, (const char*)&reuse_port, sizeof(reuse_port)) == 0;
+#else
+	int reuse_port = 1;
+	return setsockopt(d, SOL_SOCKET, SO_REUSEPORT, (const char*)&reuse_port, sizeof(reuse_port)) == 0;
```
Why define `reuse_port` again, if it's the same for both platforms?

glx22 replied:
Hmm casting int* to const char* feels weird to me, but maybe there's some magic involved.

TrueBrain replied:
@glx22: lets not look into how `setsockopt` works, you will be disappointed.

TrueBrain replied:
@spnda: I looked at this several times already, but moving it out of the `#ifdef` block looks really weird. Mostly as `setsockopt` is just misusing a variable like this in the first place. So I rather leave it "duplicated", just to make supporting platform quirks easier and make it look even less weird.

spnda commented on src/network/network.cpp:
```diff
@@ -452,6 +452,41 @@ static void CheckPauseOnJoin()
 }
 
+/**
+ * Parse the company part ("#company" postfix) of a connecting stirng.
```
Spelling issue

spnda commented on src/network/network_coordinator.cpp:
```diff
@@ -0,0 +1,701 @@
+	if (_network_dedicated) {
+		std::string connection_type;
+		switch (_network_server_connection_type) {
+			case CONNECTION_TYPE_ISOLATED: connection_type = "!! Remote connections blocked !!"; break;
```
We warn a server admin here that the remote connections are blocked. I don't think this fits together well with the message that comes just after. I think this should be a more noticeable warning message that players cannot even join the server, even though the server admin does indeed receive the message that their server is registered with the coordinator.

TrueBrain replied:
Suggestions?

spnda replied:
First thing that comes to mind is a bold warning in yellow text in a seperate message.

TrueBrain replied:
Bit the issue I had with this: normally, even for the worst connections, it falls back to TURN. So this means something is going terrible wrong. And I have a hard time getting this message across without being all technical and stuff. So I was looking for something simple and to the point, which will cause people to look for what is going on on the forums etc ..

but that message is being a pita for me, as I can't find wording that works :) So any input on the text itself would be much appreciated (not only looking at you, but at anyone :P)

embeddedt replied:
Might I suggest `None (all incoming connections blocked)` as the message? I think that gets the point across even to non-technical players.

TrueBrain replied:
That sounds a lot better! Tnx :D

The only improvement I would fancy is if we can make it clear we are not the ones doing the blocking .. but without making a very lengthy sentence, that is tricky ;)

embeddedt replied:
`None (blocked by firewall/network configuration)`?

spnda commented on src/network/network_coordinator.h:
```diff
@@ -0,0 +1,103 @@
+	std::chrono::steady_clock::time_point next_update; ///< When to send the next update (if server and public).
+	std::map<std::string, TCPServerConnecter *> connecter; ///< Based on tokens, the current connecters that are pending.
+	std::map<std::string, TCPServerConnecter *> connecter_pre; ///< Based on join-keys, the current connecters that are pending.
+	std::map<std::string, std::map<int, std::unique_ptr<ClientNetworkStunSocketHandler>>> stun_handlers; ///< All pending STUN handlers.
```
I'm not an expert on how all STUN handlers are managed and they work, but wouldn't a `std::unordered_map` be better in this case? I mean, is there really much reason for them to be ordered?

TrueBrain replied:
No, the order is not important. In fact, it always only walks all elements. But that is not what I picked `std::map`.

I simply looked at our code-base, and noticed we basically always use `std::map`, also where `std::unordered_map` wouldn't be an issue. Then I read online about the difference, and basically you have two camps. 1 yelling to use the latter as it is `O(1)`, the other saying that means little, as often the `O(1)` of `std::unordered_map` is slower than the `O(log N)` of `std::map` (and they continue on debating rb-trees vs bucket-linked-lists, ofc). Then I shrug, give up, and just do what was already done :P

Some parts of C++ are to iffy for me to have an opinion about. This is one of them. I leave it to people more knowledgeable than me about C++ to give a call on this :P

rubidium42 replied:
Conceptually `std::unordered_map` would indeed be the better choice. However, since the number of elements is tiny (after all, it is for active connection attempts to servers), the benefits of `std::unordered_map` for lookups/insertions will probably not outweigh the extra cost for hashing and maintaining the hash table compared to a `std::map` with usually 0 or 1, or maybe 2 or 3 elements.
All in all it does not really matter in this case, and then keeping it "simple" might be the better choice.

spnda replied:
I'm sorry if this is total bullshit, but theoretically the `unordered_map` would usually be faster with many entries. So if we only have a couple of entries, sure, use `std::map` as it'll be quicker on average. And I couldn't see so many socket handlers being required so I guess the minor speed boost of `unordered_map` would be negligible?
The question is, how fast the `std::hash` functions are on `std::string` and `int` to make the unordered_map actually quicker when adding/removing entries from the map.

![image](https://user-images.githubusercontent.com/43609023/123868538-20fc5f00-d930-11eb-9f49-5ff776309faa.png)

However from the following benchmark, we can see that `std::hash` on strings with over 16 characters can get very slow, up until the point of constantly being slower than a full lookup in a `std::map`. And as the keys here are 32 characters long, changing it to `std::unordered_map` would be slower anyway.

```
# 1st number column is insert time, 2nd is fetch time
 ** Integer Keys ** 
 unordered:      137      15
   ordered:      168      81
 ** Random String Keys ** 
 unordered:       55      50
   ordered:       33      31
 ** Real Words Keys ** 
 unordered:      278      76
   ordered:      516     298
   ```
   
Conclusion: With 32 character tokens as keys, the `std::unordered_map` is slower than ordinary `std::map` implementation, therefore we shouldn't change.

James103 commented:
@TrueBrain You mentioned that you load tested the Game Controller and showed that it can handle at least 10k clients. Do you also plan on load testing the TURN server as well, in order to get a good feel for how many active clients and servers it can support so that the network can scale accordingly?

TrueBrain commented:
> 
> 
> @TrueBrain You mentioned that you load tested the Game Controller and showed that it can handle at least 10k clients. Do you also plan on load testing the TURN server as well, in order to get a good feel for how many active clients and servers it can support so that the network can scale accordingly?

Yeah, tested that in one go too! Also the STUN server btw. Seems to all work and seems rather stable. Python and asyncio have a very small overhead per connected client (few KB of RAM, that is it). And CPU is the one thing we aren't short in, so not a real concern. So that all seems to work fine during testing. Of course .. real world can be a different story (stalling connections, very slow ones, etc) .. but I think we fixed most of those issues already with the current master-server (the framework used is identical between the two).

The reason for the 10k btw is that my computer couldn't spawn more clients in a useful way :P

rubidium42 commented on src/network/core/tcp_connect.cpp:
```diff
@@ -60,6 +63,14 @@ void TCPConnecter::Connect(addrinfo *address)
 
+	if (this->bind_address.GetPort() > 0) {
+		if (bind(sock, (const sockaddr *)this->bind_address.GetAddress(), this->bind_address.GetAddressLength()) != 0) {
+			Debug(net, 1, "Could not bind socket on {}: {}", this->bind_address.GetAddressAsString().c_str(), NetworkError::GetLast().AsString());
```
`.c_str()` is not needed anymore with `Debug`.

rubidium42 commented on src/network/network_coordinator.cpp:
```diff
@@ -0,0 +1,240 @@
+			if (std::chrono::steady_clock::now() > last_attempt + std::chrono::seconds(1) * last_attempt_backoff) {
+				last_attempt = std::chrono::steady_clock::now();
+				/* Delay reconnecting with up to 32 seconds. */
+				if (last_attempt_backoff < 30) {
```
The comment and the backoff value "conflict" with each other. Though functionally it will do exactly as the comment says, it's hard too actually see that. With `< 32` it would probably be clearer and less likely to cause questions. Except maybe why 32 and not 30, but that can be gleamed from the multiplication below.

rubidium42 commented on src/lang/english.txt:
```diff
@@ -2045,12 +2045,12 @@ STR_NETWORK_SERVER_LIST_SEARCH_SERVER_INTERNET_TOOLTIP          :{BLACK}Search i
 
 STR_NETWORK_SERVER_LIST_PLAYER_NAME_OSKTITLE                    :{BLACK}Enter your name
-STR_NETWORK_SERVER_LIST_ENTER_IP                                :{BLACK}Enter the address of the host
+STR_NETWORK_SERVER_LIST_ENTER_SERVER_ADDRESS                    :{BLACK}Enter the IP:port or invite-code
```
Practically it would be `address:port`, as it would still be resolving DNS names, right? Although I can imagine using IP as that might make it clearer to some people than address. Not sure what to think about it.

rubidium42 commented on src/network/core/address.cpp:
```diff
@@ -400,17 +399,40 @@ void NetworkAddress::Listen(int socktype, SocketList *sockets)
+{
+	sockaddr_storage addr = {};
+	socklen_t addr_len = sizeof(addr);
+	getpeername(sock, (sockaddr *)&addr, &addr_len);
```
Not sure whether it is likely that this function returns an error, but I think it would be better to at least log the error and return an "empty" network address than one composed of garbage data from the stack.

rubidium42 commented on src/network/core/address.cpp:
```diff
@@ -400,17 +399,40 @@ void NetworkAddress::Listen(int socktype, SocketList *sockets)
+{
+	sockaddr_storage addr = {};
+	socklen_t addr_len = sizeof(addr);
+	getsockname(sock, (sockaddr *)&addr, &addr_len);
```
Not sure whether it is likely that this function returns an error, but I think it would be better to at least log the error and return an "empty" network address than one composed of garbage data from the stack.

rubidium42 commented on src/network/network_gui.cpp:
```diff
@@ -2073,13 +2074,14 @@ struct NetworkClientListWindow : Window {
 				break;
 
-			case WID_CL_SERVER_JOIN_KEY:
+			case WID_CL_SERVER_JOIN_KEY: {
```
Does adding this here mean that all previously commits in between fail to build? That would not be very nice for a bisect session.

TrueBrain replied:
Pfft, only for MSVC users. Who cares about those anyway? :P (sometimes it is hard to put things in the right commit when you have this many .. let me try again :D)

TrueBrain commented:
Slowly we have been merging parts of this PR into master. https://github.com/OpenTTD/OpenTTD/pull/9447 is the last piece. Once that has been merged, everything in this PR (and yes, I checked) has been merged into master too.

By splitting it up, there was a huge jump in quality of what has been delivered. It really was the right call to make everything smaller, although it took significantly more work to do so. Either way, the result is that the backend is running on AWS like the rest of our infra, and that the client/server are doing what they should be .. at least for as far as I could reasonably test.

After https://github.com/OpenTTD/OpenTTD/pull/9447 is merged, the testing is up to the public .. and I am sure they will find plenty of edge-cases for us to fix :D But we will tackle them if/when they come in.

For now, I am going to close this PR. Tnx all who contributed, and feel free to test the latest nightly for STUN .. and hopefully soon TURN will hit the nightly too :D
